### PR TITLE
[1689 - Quote endpoint] New Quote endpoint for price & fee

### LINKED ISSUE #1689: Use improved fee/price API endpoint
**Description**
The backend offers now an improved version of the price and fee endpoint.

This API endpoint will make it simpler to get the information of the fee and price for the user. We don't need anymore to ask for the fee, and adjust the amounts to match that fee.

Additionally, this endpoint has some more context on who is requesting the price, which would allow the backend to decide in giving some additional discounts to the user.

See discussion here: https://gnosisinc.slack.com/archives/C01MZ4QD763/p1634905550004400

### PR DESCRIPTION
**Pull request by W3stside**
Closes #1689 

Fetches both FEE and PRICE in one endpoint, relies on different parameters.

### TESTING
_Right now, the new quote endpoint for getting PRICE is disabled, legacy price racing is being used_
1. Open the Networks tab in your browser's dev tools
2. filter by `quote`
3. select any token pair and set an amount
4. look at the `quote` post requests made to the GP backend
5. click on one and stay inside the `headers` tab
6. scroll to the bottom and check the body object being `POST`ed
7. confirm that the parameters are all filled in
8. make sure that the `validTo` param (denoted in SECONDS) converted into milliseconds (rounded down) is a valid date in the console that can be done via `Date(Math.floor(<validToParam> * 1000))` e.g `validTo` returns `1637059250` - so type in `Date(Math.floor(1637059250 * 1000))` and make sure you get a string date like: `'Tue Nov 16 2021 10:41:25 GMT+0000 (Western European Standard Time)'` and that it is in the future
9. price and quote should be working with no new issues differing from previous versions
10. errors that occur should be properly handled as before

## TESTING (LATEST)
1. you can manually change the strategy type in the console window
2. chrome console
3. `setStrategy('LEGACY')` or `setStrategy('COWSWAP')`

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -71,3 +71,6 @@ REACT_APP_PATH_REGEX_ENS="/ipfs"
 
 # Enables mock mode (default = true)
 REACT_APP_MOCK=true
+
+# Gp Api
+REACT_APP_DEFAULT_GP_API=LEGACY
```
Comment (anxolin) [inline .env]: Can we give it a more meaningful name? 

This is the price querying strategy, so sth like:
REACT_APP_DEFAULT_PRICE_STRATEGY

Values: LEGACY, COWSWAP sounds good to me. Although now im thinkining if `DEFAULT_PRICE_STRATEGY = 'best-price-apis` and `cowswap-api` are better names for our price strategies

File: src/custom/api/gnosisProtocol/api.ts [hunk 2]
```diff
@@ -266,7 +271,7 @@ async function _handleQuoteResponse(response: Response, params?: FeeQuoteParams)
       // report to sentry
       Sentry.captureException(sentryError, {
         tags: { errorType: 'getFeeQuote' },
-        contexts: { params },
+        contexts: { params: { ...params } },
```
Comment (anxolin) [inline src/custom/api/gnosisProtocol/api.ts:274]: Is this not the same as?:
```suggestion
        contexts: { params },
```

File: src/custom/api/gnosisProtocol/api.ts [hunk 3]
```diff
@@ -276,7 +281,45 @@ async function _handleQuoteResponse(response: Response, params?: FeeQuoteParams)
...
+      ? {
+          kind: OrderKind.SELL,
+          sellAmountBeforeFee: amount,
+          ...baseParams,
+        }
+      : {
+          kind: OrderKind.BUY,
+          buyAmountAfterFee: amount,
+          ...baseParams,
+        }
```
Comment (anxolin) [inline src/custom/api/gnosisProtocol/api.ts:309]: u can simplify and refactor above: kind is already a var with the right value, ...baseParams expansion are done in both cases, so the only difference is the name of the prop where you place the amount.

so make that the only difference
const amountFiled = kind === OrderKind.SELL ? 'sellAmountBeforeFee' : 'buyAmountAfterFee'

File: src/custom/hooks/useGetGpApiStatus.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import ms from 'ms.macro'
```
Comment (anxolin) [inline src/custom/hooks/useGetGpApiStatus.ts:1]: Please rename the name of this file. It's not a "Api status" . For me api status is UP, DOWN, ...

This is a price strategy.

File: src/custom/hooks/useGetGpApiStatus.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import { useState, useEffect } from 'react'
+
+export type GpQuoteStatus = 'COWSWAP' | 'LEGACY'
+// TODO: use actual API call
```
Comment (anxolin) [inline src/custom/hooks/useGetGpApiStatus.ts:7]: Can you point to an issue in the backend repo, if we want to really give them control of this and change it in the future?

File: src/custom/hooks/useGetGpApiStatus.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+export async function checkGpQuoteApiStatus(): Promise<GpQuoteStatus> {
+  return new Promise((accept) => setTimeout(() => accept('LEGACY'), 500))
```
Comment (anxolin) [inline src/custom/hooks/useGetGpApiStatus.ts:9]: Should this use the default defined in the ENV instead of hard-coding?

File: src/custom/hooks/useGetGpApiStatus.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+}
+const GP_QUOTE_STATUS_INTERVAL_TIME = ms`2 hours`
```
Comment (anxolin) [inline src/custom/hooks/useGetGpApiStatus.ts:12]: 30min would be more than enough imo

File: src/custom/hooks/useGetGpApiStatus.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  useEffect(() => {
+    console.debug('[useGetQuoteCallback::GP API Status]::', gpQuoteApiStatus)
+
+    const checkStatus = () => {
+      checkGpQuoteApiStatus()
+        .then(setGpQuoteApiStatus)
+        .catch((err: Error) => {
+          console.error('[useGetQuoteCallback::useEffect] Error getting GP quote status::', err)
+          // Fallback to LEGACY
+          setGpQuoteApiStatus('LEGACY')
```
Comment (anxolin) [inline src/custom/hooks/useGetGpApiStatus.ts:26]: again, why fallback to a hardcoded value instead of using the env?

File: src/custom/hooks/useGetGpApiStatus.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    // Create initial call on mount
+    checkStatus()
+
+    // set interval for GP_QUOTE_STATUS_INTERVAL_TIME (2 hours)
+    const intervalId = setInterval(() => {
+      checkStatus()
+    }, GP_QUOTE_STATUS_INTERVAL_TIME)
+
+    return () => clearInterval(intervalId)
+  }, [gpQuoteApiStatus])
```
Comment (anxolin) [inline src/custom/hooks/useGetGpApiStatus.ts:39]: This hook will schedule a checkStatus on every time you use it. 

I think our project could use something like:
- https://react-query.tanstack.com/
- https://swr.vercel.app

These libraries solve exactly this problems in an elegant way. i also talked with @alfetopito about this. I think this data logics are something that we reinvent the wheel every time (fetching, caching, invalidating, coordinating multiple quering of the same endpoint, loding flags, error handling, etc)

I think we should choose one of the two.

I reviewed SRW and its great, and very performant! on most comparisons they show it as a superior approach than React Query in terms of performance and production readyness. Although, React Query has some more features (and bundle size).

File: cypress-custom/support/commands.js [hunk 10]
```diff
@@ -53,7 +53,7 @@ function enterOutputAmount(tokenAddress, amount, selectToken = false) {
 }
 
 function stubResponse({ url, alias = 'stubbedResponse', body }) {
-  cy.intercept({ method: 'GET', url }, _responseHandlerFactory(body)).as(alias)
+  cy.intercept({ method: 'POST', url }, _responseHandlerFactory(body)).as(alias)
```
Comment (anxolin) [inline cypress-custom/support/commands.js:56]: why this change?

File: src/custom/api/gnosisProtocol/errors/OperatorError.ts [hunk 11]
```diff
@@ -34,6 +35,7 @@ export enum ApiErrorCodeDetails {
   InsufficientValidTo = 'The order you are signing is already expired. This can happen if you set a short expiration in the settings and waited too long before signing the transaction. Please try again.',
   InsufficientFunds = "The account doesn't have enough funds",
   UnsupportedToken = 'One of the tokens you are trading is unsupported. Please read the FAQ for more info.',
+  SellAmountDoesNotCoverFee = 'The sell amount for the sell order is lower than the fee.',
```
Comment (anxolin) [inline src/custom/api/gnosisProtocol/errors/OperatorError.ts:38]: Can you verify all our errors are in sync since you are touching this. I believe at least the NotEnoughLiquidy error  was added in the backend but not in the frontend

**Review by anxolin — COMMENTED** [review]
I will submit my review for now, but I haven't finish (since u mentioned this is still not ready)

File: src/custom/api/gnosisProtocol/api.ts (hunk 2 above)
Comment (W3stside) [reply src/custom/api/gnosisProtocol/api.ts:274]: TS was complaining

File: src/custom/api/gnosisProtocol/api.ts (hunk 3 above)
Comment (W3stside) [reply src/custom/api/gnosisProtocol/api.ts:309]: TS will complain

File: cypress-custom/support/commands.js (hunk 10 above)
Comment (W3stside) [reply cypress-custom/support/commands.js:56]: Quote endpoint is a POST

File: .env (hunk 1 above)
Comment (W3stside) [reply .env]: i actually removed this, i don't think it's necessary as an environment variable and can just be used as a constant

File: src/custom/hooks/useGetGpApiStatus.ts (hunk 9 above)
Comment (W3stside) [reply src/custom/hooks/useGetGpApiStatus.ts:39]: ok we can add that library later... definitely not for this PR

Comment (W3stside) [reply src/custom/hooks/useGetGpApiStatus.ts:39]: i can consider making it a top level updater so it's only called once

File: src/custom/hooks/useGetGpApiStatus.ts (hunk 4 above)
Comment (W3stside) [reply src/custom/hooks/useGetGpApiStatus.ts]: done

File: .env (hunk 1 above)
Comment (anxolin) [reply .env]: Fine with me too

File: src/custom/api/gnosisProtocol/api.ts (hunk 3 above)
Comment (anxolin) [reply src/custom/api/gnosisProtocol/api.ts:309]: It didn't to me https://github.com/gnosis/cowswap/pull/1847

File: src/custom/hooks/useGetGpApiStatus.ts (hunk 9 above)
Comment (anxolin) [reply src/custom/hooks/useGetGpApiStatus.ts]: yes, fair to not mix the new library with this PR. Although not too bad to use one new feature to test the new library out. Not necessarily in this PR

File: src/custom/api/gnosisProtocol/api.ts (hunk 3 above)
Comment (W3stside) [reply src/custom/api/gnosisProtocol/api.ts:309]: it did

File: src/custom/hooks/useGetGpApiStatus.ts (hunk 5 above)
Comment (W3stside) [reply src/custom/hooks/useGetGpApiStatus.ts]: https://github.com/gnosis/gp-v2-contracts/issues/904

File: cypress-custom/support/commands.js (hunk 10 above)
Comment (anxolin) [reply cypress-custom/support/commands.js:56]: well, yes, but the command is `stubResponse`.. should u add some parameter or something?

**Comment by MareenG** [comment]
Hey @W3stside,

> make sure you get a string date like: 'Tue Nov 16 2021 10:41:25 GMT+0000 (Western European Standard Time)' and that it is in the future

I tried it and I always get a time that is like 1 sec in the future. For sure it is the future but I assume it should give a time that is at least a few min in the future?

File: src/custom/api/gnosisProtocol/errors/OperatorError.ts (hunk 11 above)
Comment (W3stside) [reply src/custom/api/gnosisProtocol/errors/OperatorError.ts:38]: #1896

File: src/custom/api/gnosisProtocol/api.ts (hunk 3 above)
Comment (anxolin) [reply src/custom/api/gnosisProtocol/api.ts:309]: Damm typescript!

**Comment by W3stside** [comment]
@alfetopito pls review this change: 3b44af4

**Comment by elena-zh** [comment]
Hey @W3stside , I noticed that only in this PR swap amount is not displayed when XDAI/ETH is displayed as a sell token. Moreover, no amounts in the Wrap your XDAI/ETH section (should correspond to the amount in the sell field)
image

Besides, I can face #1964 issue in here.

**Comment by elena-zh** [comment]
Besides, I can face this error 
image

**Comment by elena-zh** [comment]
Fees are not displayed when they exceed from a sell amount (no matter if a user connected or not connected to the app)
no fees exceed
no fees exceed1

**Comment by W3stside** [comment]
> Fees are not displayed when they exceed from a sell amount (no matter if a user connected or not connected to the app) no fees exceed no fees exceed1

not relevant to this PR

**Comment by elena-zh** [comment]
> > Fees are not displayed when they exceed from a sell amount (no matter if a user connected or not connected to the app) no fees exceed no fees exceed1
> 
> not relevant to this PR

@W3stside , could you please clarify to what PR? I will need to open a new issue for this with the link to it. 
Currently, I see that in this PR `Web: v1.7.0-rc.0` app version is used. 
The issue is reproducible inside of this PR, but not in Stage or Barn with the same section
image
Barn: 
image

In Prod also everything works as expected.

**Comment by elena-zh** [comment]
Wrap your ETH/XDAI section looks good to me now

**Review by anxolin — COMMENTED** [review]
~This is creating too many requests, I think it has a re-render issue~

~If you follow the instructions, it's easy to reproduce~

fixed in 8b90b2b

**Comment by W3stside** [comment]
> This is creating too many requests, I think it has a re-render issue
> 
> If you follow the instructions, it's easy to reproduce

it's kind of expected, it's the issue i raised on slack: https://gnosisinc.slack.com/archives/C025G521XQD/p1639042676003900

the issue of validTo time for calling fee when expired

File: src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts [hunk 12]
```diff
@@ -34,14 +36,17 @@ async function _getOrderPrice(chainId: ChainId, order: Order) {
     chainId,
     amount,
     kind: order.kind,
+    sellToken: order.sellToken,
+    buyToken: order.buyToken,
     baseToken,
     quoteToken,
     fromDecimals: order.inputToken.decimals,
     toDecimals: order.outputToken.decimals,
+    validTo: Math.ceil(Date.now() / 1000 + 3000),
```
Comment (alfetopito) [inline src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts:45]: It looks like it's correct, but I have 2 suggestions:

1. Magic number 3000. What's that for, and why 3000?
2. I would add 3000 after the `ceil` method, to be more explicit. After all, you are adding an integer and there is no need for rounding that

Bonus: create a helper function for that

File: src/custom/hooks/useGetGpPriceStrategy.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
+import ms from 'ms.macro'
+import { useState, useEffect } from 'react'
+import { DEFAULT_GP_PRICE_STRATEGY } from 'constants/index'
+
+export type GpPriceStrategy = 'COWSWAP' | 'LEGACY'
+// TODO: use actual API call
+// https://github.com/gnosis/gp-v2-contracts/issues/904
+export async function checkGpPriceStrategy(): Promise<GpPriceStrategy> {
+  return new Promise((accept) => setTimeout(() => accept(DEFAULT_GP_PRICE_STRATEGY), 500))
+}
```
Comment (alfetopito) [inline src/custom/hooks/useGetGpPriceStrategy.ts:10]: What's the progress on this?
I checked the mentioned issue but it doesn't contain any progress

File: src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts (hunk 12 above)
Comment (alfetopito) [reply src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts:45]: Actually, you can use `calculateValidTo`, right?

File: src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts [hunk 14]
```diff
@@ -94,19 +101,23 @@ export function UnfillableOrdersUpdater(): null {
...
+        _getOrderPrice(chainId, order, strategy).then((quote) => {
+          if (quote) {
+            const [promisedPrice] = quote
+            const price = getPromiseFulfilledValue(promisedPrice, null)
+            console.debug(
+              `[UnfillableOrdersUpdater::updateUnfillable] did we get any price? ${order.id.slice(0, 8)}|${index}`,
+              price ? price.amount : 'no :('
+            )
+            price?.amount && updateIsUnfillableFlag(chainId, order, price)
+          }
```
Comment (alfetopito) [inline src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts:113]: Can you add a debug log when there's no `quote`?

File: src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts (hunk 12 above)
Comment (W3stside) [reply src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts:45]: order.validTo should be fine here @alfetopito

File: cypress-custom/support/commands.js (hunk 10 above)
Comment (W3stside) [reply cypress-custom/support/commands.js:56]: :+1: done

File: src/custom/hooks/useGetGpPriceStrategy.ts (hunk 13 above)
Comment (W3stside) [reply src/custom/hooks/useGetGpPriceStrategy.ts:11]: will ask backend - @nlordell have you guys been able to take a look at this? let me know if this is something i can help with

Comment (nlordell) [reply src/custom/hooks/useGetGpPriceStrategy.ts:11]: I was under the impression we decided to add this to Github configuration to make it easier to change.

Comment (nlordell) [reply src/custom/hooks/useGetGpPriceStrategy.ts:11]: Here a link to the Slack thread:
https://gnosisinc.slack.com/archives/CPZA1AGKY/p1637062028104200

**Comment by W3stside** [comment]
> > > Fees are not displayed when they exceed from a sell amount (no matter if a user connected or not connected to the app) 

related to https://gnosisinc.slack.com/archives/CPZA1AGKY/p1639652242155600

**Comment by W3stside** [comment]
despite https://github.com/gnosis/cowswap/pull/1835#issuecomment-995684422, merging this to release 1.8.0 for an RC
@gnosis/gp-frontend @gnosis/gp-qa
